packages/flutter_fuchsia/lib/src/flutter_views.dart

```diff
@@ -0,0 +1,187 @@
+/// calls.
+final HashMap<int, FuchsiaDartVm> _fuchsiaDartVmCache =
+    new HashMap<int, FuchsiaDartVm>();
+
+/// Returns a list of flutter views for the given [ipv4Address].
+/// TODO(awdavies): just returns flutter view names. Needs to return
+/// FlutterView objects of some kind that contain a JSON RPC peer.
+Future<List<String>> getFlutterViews(
+    String ipv4Address, String fuchsiaRoot, String buildType) async {
+  final List<String> views = <FlutterView>[];
+  final List<_ForwardedPort> ports =
+      await _forwardLocalPortsToDeviceServicePorts(
+          ipv4Address, fuchsiaRoot, buildType);
+  if (ports.isEmpty) {
+    return views;
+  }
+  for (_ForwardedPort fp in ports) {
+    if (!await _checkPort(fp.port)) continue;
+    final FuchsiaDartVm vmService = await _getFuchsiaDartVm(fp.port);
+    List<String> viewNames = await vmService.listFlutterViewsByName();
+    views.addAll(viewNames);
+  }
+  await Future.wait(ports.map((_ForwardedPort fp) => fp.stop()));
+  return views;
+}
+
+Future<bool> _checkPort(int port) async {
+  bool connected = true;
+  Socket s;
+  try {
+    s = await Socket.connect(ipv4Loopback, port);
+  } catch (_) {
+    connected = false;
+  }
+  if (s != null) await s.close();
+  return connected;
+}
+
+Future<FuchsiaDartVm> _getFuchsiaDartVm(int port) async {
+  if (!_fuchsiaDartVmCache.containsKey(port)) {
+    final String addr = 'http://$ipv4Loopback:$port';
+    final Uri uri = Uri.parse(addr);
+    final FuchsiaDartVm fuchsiaDartVm = FuchsiaDartVm.connect(uri);
+    _fuchsiaDartVmCache[port] = fuchsiaDartVm;
+  }
+  return _fuchsiaDartVmCache[port];
+}
+
+/// Forwards a series of local device ports to the [deviceIpv4Address] using SSH
+/// port forwarding. Returns a [List] of [_ForwardedPort] objects that the caller
+/// must close when done using. Needs [fuchsiaRoot] and [buildType] to
+/// determine the path for the SSH config.
+Future<List<_ForwardedPort>> _forwardLocalPortsToDeviceServicePorts(
+    String deviceIpv4Address, String fuchsiaRoot, String buildType) async {
+  final String config = '$fuchsiaRoot/out/$buildType/ssh-keys/ssh_config';
+  final List<int> servicePorts =
+      await getDeviceServicePorts(deviceIpv4Address, fuchsiaRoot, buildType);
+  return Future.wait(servicePorts.map((int deviceServicePort) {
+    return _ForwardedPort.start(config, deviceIpv4Address, deviceServicePort);
+  }));
+}
+
+/// Returns a list of the device service ports on success, else returns an empty
+/// list.
+Future<List<int>> getDeviceServicePorts(
+    String ipv4Address, String fuchsiaRoot, String buildType) async {
+  final FuchsiaDeviceCommandRunner runner = new FuchsiaDeviceCommandRunner(
+    ipv4Address: ipv4Address,
+    fuchsiaRoot: fuchsiaRoot,
+    buildType: buildType,
+  );
+  final List<String> lsOutput = await runner.run('ls /tmp/dart.services');
+  final List<int> ports = <int>[];
+  for (String s in lsOutput) {
+    final String trimmed = s.trim();
```

Comment:
It would be useful to provide an example string this code is trying to parse in a comment.

Reply:
good idea

Reply:
Done.